### PR TITLE
mtl: Image format properties

### LINKED ISSUE #1948: [mtl] fix the mipmap and resource size for image format properties
When running CTS on Metal, there is a lot of errors coming from `image_format_properties` tests:
> maxMipLevels is not full mip pyramid size
> maxResourceSize smaller than minimum required size

These should be easy to address in our Metal backend (and others?).

### PR DESCRIPTION
**Pull request by grovesNL**
Fixes #1948 (several CTS errors)

PR checklist:
- [X] `make` succeeds (on *nix)
- [X] `make reftests` succeeds
- [X] tested examples with the following backends: Metal

Some notes:

- Resource size still labeled as `TODO`: maximum texture size is unknown, so I've increased it for now. I think this is probably okay because it's still possible for image creation to fail (see [the note about this in the Vulkan specification](https://www.khronos.org/registry/vulkan/specs/1.1-extensions/man/html/VkImageFormatProperties.html)). I've filed a bug report with Apple to ask whether the feature set tables could be expanded to include this information. The only maximum resource size listed at the moment is for buffers, which doesn't apply to textures that aren't backed by a `MTLBuffer`, see for example [this forum post discussing this limit](https://moltengl.com/forums/topic/uploading-large-texture-data/).
- It appears we were using the wrong value for `max_levels` (it's supposed to be max mipmap levels). I don't see a value for this in the feature set table either, so until this can be clarified I simply calculated the number of expected mipmap levels using the formula in the Vulkan specification ([`maxMipLevels must be equal to ⌈log2(max(width, height, depth))⌉ + 1`](https://www.khronos.org/registry/vulkan/specs/1.1-extensions/man/html/VkImageFormatProperties.html)).
- Vulkan restricts 3D images to only support a single layer (see [valid usage for VkImageCreateInfo - `If imageType is VK_IMAGE_TYPE_3D, arrayLayers must be 1`](https://www.khronos.org/registry/vulkan/specs/1.1-extensions/man/html/VkImageCreateInfo.html)). I've added the restriction here too, although we could add that in the portability layer instead if it's preferable. I don't know what the semantics are for 3D image layered images in Metal, but it seems we don't have that option in Vulkan for now at least.

### CODE REVIEW

File: src/backend/metal/src/device.rs [hunk 1]
```diff
@@ -228,18 +228,22 @@ impl hal::PhysicalDevice<Backend> for PhysicalDevice {
...
         map_format(format).map(|_| image::FormatProperties {
-            max_extent: image::Extent {
-                width: 4096,
-                height: if dimensions >= 2 { 4096 } else { 1 },
-                depth: if dimensions >= 3 { 4096 } else { 1 },
-            },
-            max_levels: if format::Aspects::COLOR.contains(desc.aspects) { 16 } else { 1 },
-            max_layers: 2048,
+            max_extent: image::Extent { width, height, depth },
+            max_levels: max_dimension.log2().ceil() as u8 + 1,
```
Comment (kvark) [inline src/backend/metal/src/device.rs]: it is a regression from the rule that only COLOR images can have mipmaps at all

Comment (grovesNL) [reply src/backend/metal/src/device.rs]: Is there a spec reference for that? I didn't see this restriction for Vulkan though I remember OpenGL had some restrictions on depth/stencil mipmaps

Comment (kvark) [reply src/backend/metal/src/device.rs]: It's based on the Metal validation errors

**Comment by kvark** [comment]
I can't find that validation error, and neither the docs section describing this limitation.
Let's merge this now, and I'll fix it later if it becomes a problem.
bors r+

**Pull request merged**